File: JDK11/opentraceability/src/main/java/opentraceability/models/events/kdes/eventkdedouble.java
package opentraceability.models.events.kdes;

import opentraceability.Constants;
import opentraceability.interfaces.IEventKDE;
import org.json.JSONObject;
import org.w3c.dom.Document;
import org.w3c.dom.Element;

import javax.xml.parsers.DocumentBuilderFactory;
import kotlin.reflect.Type;
import kotlin.reflect.full.createType;

public class EventKDEDouble extends EventKDEBase implements IEventKDE {
    public Type valueType = Double.class.createType();
    public Double value = null;

    public EventKDEDouble(){}

    public EventKDEDouble(String ns, String name) {
        this.namespace = ns;
        this.name = name;
    }

    @Override
    public Object getJson() {
        if (this.value == null) {
            return null;
        } else {
            JSONObject json = new JSONObject();
            json.put("keyName", this.value);
            return json;
        }
    }

    @Override
    public Element getXml() {
        Double value = this.value;
        if (value == null) {
            return null;
        }

        DocumentBuilderFactory docFactory = DocumentBuilderFactory.newInstance();
        Document document;
        try {
            document = docFactory.newDocumentBuilder().newDocument();
        } catch (Exception ex) {
            ex.printStackTrace();
            return null;
        }

        Element element = document.createElementNS(this.namespace, this.name);
        element.setTextContent(value.toString());

        Element xsiTypeAttr = document.createAttributeNS(Constants.XSI_NAMESPACE, "type");
        xsiTypeAttr.setAttributeNS(Constants.XSI_NAMESPACE, "value", "number");
        element.setAttributeNodeNS(xsiTypeAttr);

        return element;
    }

    @Override
    public void setFromJson(JSONObject json) {
        this.value = json.getDouble("value");
    }

    @Override
    public void setFromXml(Element xml) {
        try {
            this.value = Double.parseDouble(xml.getTextContent());
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    @Override
    public String toString() {
        return value == null ? "" : value.toString();
    }
}

File: Java/OpenTraceability/src/main/opentraceability/models/events/kdes/eventkdedouble.kt
package models.events.kdes

import com.fasterxml.jackson.core.JsonToken
import interfaces.IEventKDE
import org.json.simple.JSONObject
import java.lang.reflect.Type
import javax.xml.parsers.DocumentBuilderFactory
import org.w3c.dom.Element
import org.w3c.dom.Document

class EventKDEDouble: EventKDEBase, IEventKDE {

    override var ValueType: Type = Double::class.java

    var Value: Double? = null

    constructor() {

    }

    constructor(ns: String, name: String) {
        this.Namespace = ns;
        this.Name = name;
    }


    override fun GetJson(): Any? {
        if (this.Value == null) {
            return null
        } else {
            val json = JSONObject()
            json.put("keyName", this.Value)
            return json
        }
    }

    override fun GetXml(): Element? {
        val value = this.Value ?: return null

        val docFactory = DocumentBuilderFactory.newInstance()
        val docBuilder = docFactory.newDocumentBuilder()
        val document: Document = docBuilder.newDocument()

        val element = document.createElementNS(this.Namespace, this.Name)
        element.textContent = value.toString()

        val xsiTypeAttr = document.createAttributeNS(Constants.XSI_NAMESPACE, "type")
        xsiTypeAttr.value = "number"
        element.setAttributeNodeNS(xsiTypeAttr)

        return element
    }

    override fun SetFromJson(json: JSONObject) {
        this.Value = json["value"] as Double
    }

    override fun SetFromXml(xml: Element) {
        this.Value = xml.textContent.toDoubleOrNull()
    }

    override fun toString(): String {
        if (Value != null) {
            return Value.toString()
        } else {
            return ""
        }
    }
}
